src/devices/Ht1632/Ht1632.cs

```diff
@@ -0,0 +1,189 @@
+            { 5, Command.PWM_Duty_5 },
+            { 6, Command.PWM_Duty_6 },
+            { 7, Command.PWM_Duty_7 },
+            { 8, Command.PWM_Duty_8 },
+            { 9, Command.PWM_Duty_9 },
+            { 10, Command.PWM_Duty_10 },
+            { 11, Command.PWM_Duty_11 },
+            { 12, Command.PWM_Duty_12 },
+            { 13, Command.PWM_Duty_13 },
+            { 14, Command.PWM_Duty_14 },
+            { 15, Command.PWM_Duty_15 },
+            { 16, Command.PWM_Duty_16 },
+        };
+
+        private readonly Ht1632PinMapping _pinMapping;
+        private readonly int _cs;
+        private readonly int _wr;
+        private readonly int _data;
+        private readonly bool _shouldDispose;
+
+        private GpioController? _controller;
+
+        /// <summary>
+        /// HT1632C 32x8 and 24x16 LED Driver
+        /// </summary>
+        /// <param name="pinMapping">The pin mapping to use by the binding.</param>
+        /// <param name="gpioController">The GPIO Controller used for interrupt handling.</param>
+        /// <param name="shouldDispose">True (the default) if the GPIO controller shall be disposed when disposing this instance.</param>
+        public Ht1632(Ht1632PinMapping pinMapping, GpioController? gpioController = null, bool shouldDispose = true)
+        {
+            _shouldDispose = shouldDispose || gpioController is null;
+            _controller = gpioController ?? new GpioController();
+            _pinMapping = pinMapping;
+            _cs = _pinMapping.CS;
+            _wr = _pinMapping.WR;
+            _data = _pinMapping.DATA;
+
+            SetupPins();
+        }
+
+        /// <summary>
+        /// Command Mode
+        /// </summary>
+        /// <param name="commands">Commands to sent</param>
+        public void WriteCommand(params Command[] commands)
+        {
+            Begin();
+
+            WriteBits((byte)Id.COMMAND, 3);
+
+            foreach (Command command in commands)
+            {
+                WriteBits((byte)(command), 8);
+                WriteBits(0, 1);
+            }
+
+            End();
+        }
+
+        /// <summary>
+        /// WRITE Mode - Successive Address Writing
+        /// </summary>
+        /// <param name="address">Memory Address (MA) - 0b00_A6_A5_A4_A3_A2_A1_A0</param>
+        /// <param name="data">Data (MA, MA+1, ...) - 0b0000_D0_D1_D2_D3</param>
+        public void WriteData(byte address, params byte[] data)
+        {
+            Begin();
+
+            WriteBits((byte)Id.WRITE, 3);
+            WriteBits(address, 7);
+
+            foreach (byte d in data)
+            {
+                WriteBits(d, 4);
+            }
+
+            End();
+        }
+
+        /// <inheritdoc/>
+        public void Dispose()
+        {
+            if (_shouldDispose)
+            {
+                _controller?.Dispose();
+                _controller = null;
+            }
+            else
+            {
+                _controller?.ClosePin(_cs);
```

Comment:
in theory to have it super clean, you're supposed to close also the pins even if you don't dispose it. So that can be common.
To avoid double disposing, you can use the _isDisposed pattern or equivalent as well.